text/0000-dockerfiles.md

```diff
@@ -27,25 +27,65 @@ This RFC proposes that we replace stackpacks with multi-purpose build-time and r
 
-A buildpack app may have a build.Dockerfile and/or run.Dockerfile in its app directory. A run.Dockerfile is applied to the selected runtime base image after the detection phase. A build.Dockerfile is applied to the build-time base image before the detection phase.
+A builder may specify any number of executable "hooks" in `/cnb/hooks.d/`.
+Hook files in this directory are executed in the context of app directory and must output Dockerfiles that are applied to the build-time and runtime base images before an image is built.
 
-Both Dockerfiles must accept `base_image` and `build_id` args. The `base_image` arg allows the lifecycle to specify the original base image. The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`.  
+If a hook exits with a non-zero status value, the build fails. If a hook exits with a zero status value and no output, the hook is ignored. Directories and non-executable files are ignored.
+
+Each executable must be in the format `/cnb/hooks.d/<name>.(build.|run.|)<format>`, where build, run, or empty specify the build-time base image, runtime base image, or both bash images, respectively. The only valid format is `Dockerfile`, although support for, e.g. LLB JSON, could be added in the future. 
+
+A runtime Dockerfile is applied to the selected runtime base image after the detection phase.
+A build-time Dockerfile is applied to the build-time base image before the detection phase.
+
+Both Dockerfiles must accept `base_image` and `build_id` args.
+The `base_image` arg allows the lifecycle to specify the original base image.
+The `build_id` arg allows the app developer to bust the cache after a certain layer and must be defaulted to `0`. When the `$build_id` arg is referenced in a `RUN` instruction, all subsequent layerrs will be rebuilt on the next build (as the value will change).
 
 A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of app-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before `run.Dockerfile` is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.
 
+
+#### Example: App-specified Dockerfile Hook
+
+This example hook would allow an app to provide runtime and build-time base image extensions as "run.Dockerfile" and "build.Dockerfile."
+The app developer can decide whether the extensions are rebasable.
+
+##### `/cnb/hooks.d/app.build.Dockerfile`
+```
+#!/bin/sh
+cat build.Dockerfile
+```
+##### `/cnb/hooks.d/app.run.Dockerfile`
+```
+#!/bin/sh
+cat run.Dockerfile
+```
+
+#### Example: RPM Dockerfile Hook
+
+This example hook would allow a builder to install RPMs for each language runtime.
+
+Note: The Dockerfiles referenced must disable rebasing, and build times will be slower compared to buildpack-provided runtimes.
```

Comment:
Can you at the beginning of this RFC describe clearly if rebasing will be supported or not when dockerfiles have been provided by an application (= developer project) or by the platform ?

If run images cannot be rebased, what will be then in this case the workaround (if it exists) as rebasing is a key feature ?

Reply:
As mentioned in the first section:
> A runtime base image may indicate that it preserves ABI compatibility by adding the label `io.buildpacks.rebasable=true`. In the case of builder-specified Dockerfiles, `io.buildpacks.rebasable=false` is set automatically before a runtime Dockerfile is applied and must be explicitly set to `true` if desired. Rebasing an app without this label set to `true` requires passing a new `--force` flag to `pack rebase`.

So it's up to the Dockerfile author to enable rebasing. That said, Dockerfiles installing OS packages (like RPMs) should never allow rebasing, because it is unsafe. There is no workaround for this other than not using OS packages at build-time. OS packages are arbitrary functions that can modify existing files.

Reply:
> OS packages are arbitrary functions that can modify existing files.

Is that the only criterium for knowing if a Dockerfile would be rebasable? So if you knew for sure that your RPM didn't change any files but that it only added files you could safely set the rebasable flag to true?

Reply:
If installing your RPM (and all dependencies of that RPM) only adds net-new files, and if those files are not present in the base image (or any future base image), then it's likely safe to mark the image as rebasable.

That said, unless you unpack the RPM manually, installing your RPM modifies the package DB in /var/lib/rpm. Rebase would still be unsafe and could also, e.g., cause security scanners to pick up the wrong package versions after a rebase.

Reply:
Marking as resolved, but please feel free to reopen if needed.